core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c

```diff
@@ -969,9 +1006,49 @@ wasi_poll_oneoff(wasm_exec_env_t exec_env, const wasi_subscription_t *in,
         || !validate_native_addr(nevents_app, sizeof(uint32)))
         return (wasi_errno_t)-1;
 
-    err = wasmtime_ssp_poll_oneoff(curfds, in, out, nsubscriptions, &nevents);
-    if (err)
-        return err;
+    __wasi_timestamp_t elapsed = 0;
+    const __wasi_timestamp_t timeout = get_timeout_for_poll_oneoff(
+                                 in, nsubscriptions),
+                             time_quant = 1e9;
+    const uint64 size_to_copy = nsubscriptions * sizeof(wasi_subscription_t);
+    __wasi_subscription_t *in_copy = NULL;
+    if (size_to_copy >= UINT32_MAX
+        || (size_to_copy
+            && !(in_copy = (__wasi_subscription_t *)wasm_runtime_malloc(
+                     (uint32)size_to_copy)))) {
+        return (wasi_errno_t)-1;
+    }
+    bh_memcpy_s(in_copy, size_to_copy, in, size_to_copy);
+
+    while (1) {
+        elapsed += time_quant;
+        if (timeout != (__wasi_timestamp_t)-1 && elapsed > timeout)
+            break;
+
+        /* update timeout for clock subscription events */
+        update_clock_subscription_data(in_copy, nsubscriptions, time_quant);
```

Comment:
This should not be `time_quant`; if the timeout e.g. is 2.5s, we should pass 1s for the first 2 iterations, but for the last iteration it should be 0.5s.

Reply:
good catch, we can instead pass `min(time_quant, timeout - elapsed)`